qiskit/circuit/controlflow/while_loop.py

```diff
@@ -0,0 +1,109 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"Circuit operation representing a ``while`` loop."
+
+from typing import Optional, Tuple, Union
+
+from qiskit.circuit import Clbit, ClassicalRegister, QuantumCircuit
+from qiskit.circuit.exceptions import CircuitError
+from .control_flow import ControlFlowOp
+
+
+class WhileLoopOp(ControlFlowOp):
+    """A circuit operation which repeatedly executes a subcircuit (``body``) until
+    a condition (``condition``) evaluates as False.
+
+    Parameters:
+        condition: A condition to be checked prior to executing ``body``. Can be
+            specified as either a tuple of a ``ClassicalRegister`` to be tested
+            for equality with a given ``int``, or as a tuple of a ``Clbit`` to
+            be compared to either a ``bool`` or an ``int``.
+        body: The loop body to be repeatedly executed.
+        label: An optional label for identifying the instruction.
+
+    The classical bits used in ``condition`` must be a subset of those attached
+    to ``body``.
+
+    **Circuit symbol:**
+
+    .. parsed-literal::
+
+             ┌─────────────┐
+        q_0: ┤0            ├
+             │             │
+        q_1: ┤1            ├
+             │  while_loop │
+        q_2: ┤2            ├
+             │             │
+        c_0: ╡0            ╞
+             └─────────────┘
+
+    """
+
+    def __init__(
+        self,
+        condition: Union[
+            Tuple[ClassicalRegister, int],
+            Tuple[Clbit, int],
+            Tuple[Clbit, bool],
+        ],
```

Comment:
I think it would be better to define a condition/expression type from the get-go. Otherwise, we will have an explosion of types supported by this constructor.